src/Catalyst.KBucket/KBucket.cs

```diff
@@ -0,0 +1,503 @@
+        /// <param name="contact">
+        ///   The <see cref="IContact"/> to validate.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        ///   When <paramref name="contact"/> is null or its <see cref="IContact.Id"/>
+        ///   is null or empty.
+        /// </exception>
+        private void Validate(T contact)
+        {
+            if (contact == null) throw new ArgumentNullException("contact");
+
+            if (contact.Id == null || contact.Id.Length == 0) throw new ArgumentNullException("contact.Id");
+        }
+
+        /// <summary>
+        ///   Add the contact.
+        /// </summary>
+        /// <param name="contact"></param>
+        /// <param name="ping"></param>
+        /// <returns>
+        ///   <b>true</b> if the <paramref name="contact"/> was added; otherwise, 
+        ///   <b>false</b> and a <see cref="Ping"/> event should be raised.
+        /// </returns>
+        private bool _Add(T contact, out PingEventArgs<T> ping)
+        {
+            ping = null;
+
+            var bitIndex = 0;
+            var node = Root;
+
+            while (node.Contacts == null)
+
+                // this is not a leaf node but an inner node with 'low' and 'high'
+                // branches; we will check the appropriate bit of the identifier and
+                // delegate to the appropriate node for further processing
+                node = _DetermineNode(node, contact.Id, bitIndex++);
+
+            // check if the contact already exists
+            var index = node.IndexOf(contact.Id);
+            if (0 <= index)
+            {
+                _Update(node, index, contact);
+                return true;
+            }
+
+            if (node.Contacts.Count < ContactsPerBucket)
+            {
+                node.Contacts.Add(contact);
+                return true;
+            }
+
+            // the bucket is full
+            if (node.DontSplit)
+            {
+                // we are not allowed to split the bucket
+                // we need to ping the first this.numberOfNodesToPing
+                // in order to determine if they are alive
+                // only if one of the pinged nodes does not respond, can the new contact
+                // be added (this prevents DoS flodding with new invalid contacts)
+
+                ping = new PingEventArgs<T>
+                {
+                    Oldest = node.Contacts.Take(ContactsToPing).ToArray(),
+                    Newest = contact
+                };
+                return false;
+            }
+
+            _Split(node, bitIndex);
+            return _Add(contact, out ping);
+        }
+
+        /// <summary>
+        ///   Splits the node, redistributes contacts to the new nodes, and marks the
+        ///   node that was split as an inner node of the binary tree of nodes by
+        ///   setting this.root.contacts = null
+        /// </summary>
+        private void _Split(Bucket<T> node, int bitIndex)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Rename method '_Split' to match pascal case naming rules, trim underscores from the name.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)